deps/rabbitmq_peer_discovery_k8s/src/rabbit_peer_discovery_k8s.erl

```diff
@@ -2,248 +2,52 @@
-    ok = application:ensure_started(inets),
-    %% we cannot start this plugin yet since it depends on the rabbit app,
-    %% which is in the process of being started by the time this function is called
-    _ = application:load(rabbitmq_peer_discovery_common),
-    rabbit_peer_discovery_httpc:maybe_configure_proxy(),
-    rabbit_peer_discovery_httpc:maybe_configure_inet6().
-
--spec list_nodes() -> {ok, {Nodes :: list(), NodeType :: rabbit_types:node_type()}} | {error, Reason :: string()}.
+-spec list_nodes() -> {ok, {Nodes :: [node()] | node(), NodeType :: rabbit_types:node_type()}} | {error, Reason :: string()}.
 
 list_nodes() ->
-  case make_request() of
-    {ok, Response} ->
-      Addresses = extract_node_list(Response),
-      Nodes = lists:map(fun node_name/1, Addresses),
-      {ok, {Nodes, disc}};
-    {error, Reason} ->
-      Details = io_lib:format("Failed to fetch a list of nodes from Kubernetes API: ~ts", [Reason]),
-      rabbit_log:error(Details),
-      _ = send_event("Warning", "Failed", Details),
-      {error, Reason}
-  end.
-
--spec supports_registration() -> boolean().
-
-supports_registration() ->
-    true. %% to send event in post_registration/0
-
--spec post_registration() -> ok | {error, Reason :: string()}.
-post_registration() ->
-    Details = io_lib:format("Node ~ts is registered", [node()]),
-    send_event("Normal", "Created", Details).
-
--spec register() -> ok.
-register() ->
-    ok.
-
--spec unregister() -> ok.
-unregister() ->
-    ok.
-
--spec lock(Nodes :: [node()]) ->
-    {ok, {{ResourceId :: string(), LockRequesterId :: node()}, Nodes :: [node()]}} |
-    {error, Reason :: string()}.
-
-lock(Nodes) ->
-    Node = node(),
-    case lists:member(Node, Nodes) of
-        true ->
-            rabbit_log:info("Will try to lock connecting to nodes ~tp", [Nodes]),
-            LockId = rabbit_nodes:lock_id(Node),
-            Retries = rabbit_nodes:lock_retries(),
-            case global:set_lock(LockId, Nodes, Retries) of
-                true ->
-                    {ok, {LockId, Nodes}};
-                false ->
-                    {error, io_lib:format("Acquiring lock taking too long, bailing out after ~b retries", [Retries])}
-            end;
-        false ->
-            %% Don't try to acquire the global lock when local node is not discoverable by peers.
-            %% This branch is just an additional safety check. We should never run into this branch
-            %% because the local Pod is in state 'Running' and we listed both ready and not-ready addresses.
-            {error, lists:flatten(io_lib:format("Local node ~ts is not part of discovered nodes ~tp", [Node, Nodes]))}
+    Nodename = atom_to_list(?MODULE:node()),
+    try
+        [[], Prefix, StatefulSetName, MyPodId, Domain] = re:split(
+                                                           Nodename,
+                                                           "([^@]+@)([^.]*-)([0-9]+)",
```

Comment:
Shouldn’t the regex include `^...$` to make sure it matches against the entire string?

Reply:
You can also use the `{capture, all_but_first}` option to get only what you want.

Reply:
The remaining part of the regexp captures the `Domain`, so no, at least not this way. I've only added `^` to make it more strict. Alternatively, there could be an explicit capture group for the domain (but it would just capture "everything else" anyway)

Reply:
Do k8s environments always use long node names?

Reply:
At least in practice - I'd say yes. Both the Operator and Bitnami's chart set it unconditionally:
https://github.com/rabbitmq/cluster-operator/blob/main/internal/resource/statefulset.go#L594-L595
https://github.com/bitnami/charts/blob/main/bitnami/rabbitmq/templates/statefulset.yaml#L251-L252

These two deployment methods should cover a vast majority of RabbitMQ on Kubernetes users.

Nevertheless, the plan is to merge this PR soon and reach out to the community to test the 4.1 alpha and confirm it works with strange DNS settings and whatever else crazy Kubernetes configurations there can be.